src/status_im/ui/screens/browser/events.cljs

```diff
@@ -10,78 +9,95 @@
 
-(defn get-new-browser [browser now]
-  (cond-> browser
-    true
-    (assoc :timestamp now)
-    (not (:browser-id browser))
-    (assoc :browser-id (random/id))
-    (not (:name browser))
-    (assoc :name (i18n/label :t/browser))
-    (:url browser)
-    (update :url (comp js/decodeURI http/normalize-url))))
-
-(defn add-browser-fx [{:keys [db now]} browser]
-  (let [new-browser (get-new-browser browser now)]
-    {:db            (update-in db [:browser/browsers (:browser-id new-browser)]
-                               merge new-browser)
-     :data-store/tx [(browser-store/save-browser-tx new-browser)]}))
+(defn update-browser-fx [{:keys [db now]} browser]
+  (let [updated-browser (assoc browser :timestamp now)]
+    {:db            (update-in db [:browser/browsers (:browser-id updated-browser)]
+                               merge updated-browser)
+     :data-store/tx [(browser-store/save-browser-tx updated-browser)]}))
+
+(defn update-browser-history-fx [cofx browser url loading]
+  (cond
+    loading
+    nil
+
+    (get-in cofx [:db :browser/options :dont-store-history-on-nav-change?])
+    (assoc-in (update-browser-fx cofx browser)
+              [:db :browser/options :dont-store-history-on-nav-change?]
+              false)
+
+    :else
+    (let [history-index (:history-index browser)
+          history       (:history browser)
+          history-url   (try (nth history history-index) (catch js/Error _))
+          new-history   (if (not= history-url url)
+                          (conj (subvec history 0 (inc history-index))
+                                url)
+                          history)
+          new-index     (dec (count new-history))]
+      (update-browser-fx cofx
+                         (assoc browser :history new-history :history-index new-index)))))
+
+(defn update-browser-and-navigate [cofx browser]
+  (merge (update-browser-fx cofx browser)
+         {:dispatch [:navigate-to :browser (:browser-id browser)]}))
 
 (handlers/register-handler-fx
  :open-dapp-in-browser
  [re-frame/trim-v]
  (fn [cofx [{:keys [name dapp-url]}]]
-   (let [browser {:browser-id name
-                  :name       name
-                  :dapp?      true
-                  :url        dapp-url}]
-     (merge (add-browser-fx cofx browser)
-            {:dispatch [:navigate-to :browser {:browser/browser-id (:browser-id browser)}]}))))
+   (let [browser {:browser-id    name
+                  :name          name
+                  :dapp?         true
+                  :history-index 0
+                  :history       [(http/normalize-and-decode-url dapp-url)]}]
+     (update-browser-and-navigate cofx browser))))
+
+(handlers/register-handler-fx
+ :open-url-in-browser
+ [re-frame/trim-v]
+ (fn [cofx [url]]
+   (let [browser {:browser-id    (random/id)
+                  :name          (i18n/label :t/browser)
+                  :history-index 0
+                  :history       [(http/normalize-and-decode-url url)]}]
```

Comment:
Some parts in common with code above. Introduce a fn?

Reply:
do you mean for creating new browser map?

Reply:
Right